File: 239-sliding-window-maximum/239-sliding-window-maximum.swift
class Solution {
    func maxSlidingWindow(_ nums: [Int], _ k: Int) -> [Int] {
        
        var deque = [Int]()
        var ans = [Int]()
        for index in 0..<k-1{
            if !deque.isEmpty && deque.first! <= index - k{
                deque.removeFirst()
            }
            
            while( !deque.isEmpty && nums[deque.last!] < nums[index]){
                deque.removeLast()
            }
            
            deque.append(index)
        }
        for index in (k-1)..<nums.count{
            
            if !deque.isEmpty && deque.first! <= index - k{
                deque.removeFirst()
            }
            
            while( !deque.isEmpty && nums[deque.last!] < nums[index]){
                deque.removeLast()
            }
            
            deque.append(index)
            ans.append(nums[deque.first!])
            
        }
        
        return ans
        
    }
}

File: 239-sliding-window-maximum/239-sliding-window-maximum.cpp
class Solution {
public:
    void insertElement(int index, deque<int> &dq, vector<int>& nums, int k ){
        while(!dq.empty() && dq.back() == (index - k )){
            dq.pop_back();
        }
        
        while( !dq.empty() && (nums[index] > nums[dq.front()] )){
            dq.pop_front();
        }
        dq.push_front(index);
    }
    
    vector<int> maxSlidingWindow(vector<int>& nums, int k) {
        vector<int> ans;
        deque<int> dq;
        
        for(int index = 0; index < k - 1; index ++ ){
            insertElement(index, dq, nums, k);
        }
        
        int size = nums.size();
        
        for(int index = k - 1; index < size; index ++ ){
            
            insertElement(index, dq, nums, k);
            
            ans.push_back(nums[dq.back()]);
        }
        
        return ans;
    }
};


// 3 

// 2 7 

 // 6 7